Add entity validation for clickhouse functions names

MeredithAnya opened the pull request:
**Current Function Validation Status:**
when a snql query comes to snuba, it goes through a lot of processing that happens in [`parse_snql_query`](https://github.com/getsentry/snuba/blob/master/snuba/query/snql/parser.py#L1295). This is done as a series of calls to [`_post_process`](https://github.com/getsentry/snuba/blob/master/snuba/query/snql/parser.py#L1266), with different processors being passed as an argument - last of which are the `VALIDATORS`.
(`VALIDATORS` is a list of two functions: `validate_query` and `validate_entities_with_query`)
https://github.com/getsentry/snuba/blob/6bc208abe7b1d05023e9af1922628bed87c8a84b/snuba/query/snql/parser.py#L1318-L1319 

The `validate_query` function is what ultimately calls the `FunctionCallsValidator`. And since `_post_process` gets called recursively, this function does.
https://github.com/getsentry/snuba/blob/6bc208abe7b1d05023e9af1922628bed87c8a84b/snuba/query/parser/validation/__init__.py#L29-L39

> **what _currently_ gets validated ?**
Well in theory each entity could have its own function validators that come from [`get_function_call_validators`](https://github.com/getsentry/snuba/blob/08cd891055c736fa964bf6f1c510004048e74e9f/snuba/datasets/entity.py#L95), but afaict this always returns `{}`. So instead we are left with just two [`default_validators`](https://github.com/getsentry/snuba/blob/08cd891055c736fa964bf6f1c510004048e74e9f/snuba/datasets/entity.py#L95). However, these validators are for validating that the function is well formed. There is no validation if the function should be allowed or not.

**Problem**
Because we are not really validating any functions beyond the two specified in the `default_validators`, it's possible that any ClickHouse function could be used, which is not great for security reasons. ClickHouse has a lot of functions (categorized in 'regular' [functions](https://github.com/getsentry/snuba/blob/08cd891055c736fa964bf6f1c510004048e74e9f/snuba/datasets/entity.py#L95), [aggregate functions](https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/) and [table functions](https://clickhouse.tech/docs/en/sql-reference/table-functions/)) 

**Goal for This PR**
We want to have basically an allow list of valid functions for entities that we can check against to make sure no one is using something we don't want. Easier to add incrementally to an allowed/valid list versus having to keep up with a deny list if ClickHouse introduces a new table function, for example.

MeredithAnya commented on snuba/query/functions.py:
```diff
@@ -102,3 +102,27 @@
+
+# Categorized based on ClickHouse docs
+# https://clickhouse.tech/docs/en/sql-reference/functions/
+REGULAR_FUNCTIONS = {
```
not sure yet if this the full list, might be missing some that we use/allow

fpacifici replied:
The safest way to verify that the list is complete, is to start the validator in dry-run mode.
This means the valdiator will perform the validation, but, if an invalid function is found, instead of rejecting the query it will log a warning and allow it, so we can find what we missed and fix the list. 
Once we do not log anything anymore we are done and we enable the actual validation.

MeredithAnya replied:
ah yeah okay the wrong part I think I meant `_parse_datetime_literals` maybe? either way, i dont think what I have now validates things like `toDateTime`

evanh replied:
We should allow type conversion functions. The _parse_datetime_literals is a specific case of `toDateTime`. In SnQL we mark datetime literals by wrapping them in `toDateTime` so we can tell the difference between a string literal that looks like a date and an actual date. Specifically the `tuple` function we use quite a bit.

There are definitely some missing things here. `has`, `if`, `ifNull`, `multiIf`. Take a look at `https://github.com/getsentry/sentry/blob/master/src/sentry/search/events/fields.py#L1584`. These are all the functions that Discover/performance sends. Most of them are aggregate functions but there are other ones as well.

MeredithAnya commented on snuba/datasets/entity.py:
```diff
@@ -92,6 +93,13 @@ def get_all_storages(self) -> Sequence[Storage]:
         """
         return self.__storages
 
+    def get_valid_functions(self) -> Set[str]:
```
seemed to make the most sense to me to have checking if the function is a valid function be separate from the actual validator of the function itself.

MeredithAnya commented on snuba/query/parser/validation/functions.py:
```diff
@@ -40,8 +41,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
+                # todo(meredith): add some metrics or logging
+                # to audit what functions are actually being used
+                # that we might have missed, later raise an err
+                print(f"{exp.function_name} is not valid for {entity}")
```
This was just for me testing, will not actually print this out in prod lol

MeredithAnya commented on snuba/datasets/entity.py:
```diff
@@ -92,6 +93,13 @@ def get_all_storages(self) -> Sequence[Storage]:
         """
         return self.__storages
 
+    def get_valid_function_names(self) -> Set[str]:
```
seemed to make the most sense to me to have checking if the function name is a valid function name be separate from the actual validator of the function itself.

evanh commented on snuba/query/parser/validation/functions.py:
```diff
@@ -40,8 +41,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
+            data_model = data_source.get_columns()
+
+        elif isinstance(data_source, ProcessableQuery):
+            entity = get_entity(data_source.get_from_clause().key)
```
You don't need to do this, `_post_process` will already recurse on the query. If you have this here you're going to validate twice.

fpacifici replied:
I think this is required actually, but not in this form `isinstance(data_source, ProcessableQuery)` is not broad enoguh.
In a hypothetical query with a hypothetical function call like
```
MATCH (
     MATCH entity
     SELECT .......whatever ....
)
SELECT shutdown_system_and_format_hdd(), event_id
```

We do want to validate the shutdown function call against the list of allowed/denied functions.
In that case data_source would be a `CompositeQuery` if I get this right.

evanh commented on snuba/query/parser/validation/functions.py:
```diff
@@ -40,8 +41,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
+        else:
+            return
+
+        for entity in entities:
```
This is correct if we had entity specific whitelists. Since we don't, I think we should make one call to check against the global list and move on.

evanh reviewed (commented):
Definitely on the right track here, well done.

The one structural thing I would change is checking the global list from the context of the entity. Once we have entity specific whitelists what you're doing would make sense.

On that note, there is something entity specific that could be checked here: some entities allow custom functions using the `CustomFunction` processor. For example here in the transactions entity. https://github.com/getsentry/snuba/blob/master/snuba/datasets/entities/transactions.py#L144.

These processors allow Sentry to pass functions that don't exist in Clickhouse (`apdex(x, y)`) and the processor will convert it into the appropriate Clickhouse function. These processors are run after your validation, so there will be terms like `FunctionCall(None, function_name="apdex"...)` that will fail your validation because `apdex` is not whitelisted and the expression hasn't been converted to a Clickhouse function yet.

fpacifici commented on snuba/query/parser/validation/functions.py:
```diff
@@ -40,8 +41,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
+
+        for entity in entities:
+            valid_funcs = entity.get_valid_function_names()
+            if exp.function_name not in valid_funcs:
+                # todo(meredith): add some metrics or logging
+                # to audit what functions are actually being used
+                # that we might have missed, later raise an err
```
I think there is something that does not look right in this approach. Specifically in the fact that the class `FunctionCallsValidator` contain the specific logic of the individual validator (validation of the function name).
The original design of this class was that it would be executed on each and every expression in the query, it would gather whatever it needed to perform the actually validation and then delegate the specific validation logic to individual `FunctionCallValidator` classes. These ones would contain the validation logic. 
This way we can have a speartion of concern and have a single responsibility in each of these classes:
- `FunctionCallsValidator` (this class) finds the list of validation rules we need for a given expression
- Each `FunctionCallValidator` object applies one and only one validation rule.

The advantage of this is that we don't pollute the code of this class with a lot of unrelated validators and each specific function call validator rule can be unit tested in isolation thoroughly.

So in order to follow that principle we should have this class relying on a `FunctionCallValidator` (let's call it AllowedFunctionValidator) to actually do the check.

Unfortunately this design needs to change to accomplish that goal as the FunctionCallValidator only takes function parameters and not the name.

What about changing the interface of FunctionCallValidator to take the name as well together with what it needs to validate the function. (This code came before entities, and it seem we may not need to pass the data model to the validator, we could pass the entity of interest).

You may pass to the `FunctionCallValidator` these fields:
- function name
- parameters 
- entities involved.

The `AllowedFunctionValidator` would have whatever it needs to fetch the list of allowed functions.
The `SignatureValidator`, having the entity, would have what it needs to fetch the data model.

Now this class does not have to know details (like finding the data model) about how the validation is done, it jsut bothers to find the right validators to apply to a given expression.

How does it sound ?

fpacifici commented on snuba/query/parser/validation/functions.py:
```diff
@@ -54,23 +85,16 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
+            validators = ChainMap(default_validators, entity_validators)
+            try:
+                # TODO: Decide whether these validators should exist at the Dataset or Entity level
+                validator = validators.get(exp.function_name)
```
IF you end up applying my suggestion above, this one would have to change a bit as the `AllowedFunctionValidator` would have to be applied to all function calls.

fpacifici reviewed (commented):
Overall seems on the good track, though I think there is some refactoring we should do to `FunctionCallsValidator` to make the design of that class work properly with composite and joined queries. They did not exist when the class was first introduced as well as entities.

MeredithAnya commented:
[UPDATES]:
Re: feedback in https://github.com/getsentry/snuba/pull/2044#pullrequestreview-723697618
* removed the global functions from the context of the entities, added `is_valid_global_function` instead that can be imported

Re: feedback in https://github.com/getsentry/snuba/pull/2044#discussion_r683891283
* made a separate `AllowedFunctionValidator`
* update the base `FunctionCallValidator` to take in the `func_name`

MeredithAnya commented on snuba/query/processors/handled_functions.py:
```diff
@@ -41,7 +41,7 @@ def __init__(self, column: str, columnset: ColumnSet):
         validator = SignatureValidator([])
         try:
-            validator.validate(exp.parameters, self.__columnset)
+            validator.validate(exp.function_name, exp.parameters, self.__columnset)
```
so part of the reason I didn't pass in the entity was because this `SignatureValidator` gets used in a lot of places and would need to be updated if we starting using the entity instead of the ColumnSet

fpacifici replied:
You should have easy access to the entity in every place where the SignatureValidator is used.

```
(.venv) ~/snuba  (feat/replacement_load_balancer *)$ git grep SignatureValidator
snuba/query/parser/validation/functions.py:from snuba.query.validation.signature import Any, Column, SignatureValidator
snuba/query/parser/validation/functions.py:    "like": SignatureValidator([Column({Array, String}), Any()]),
snuba/query/parser/validation/functions.py:    "notLike": SignatureValidator([Column({Array, String}), Any()]),
snuba/query/processors/custom_function.py:from snuba.query.validation.signature import ParamType, SignatureValidator
snuba/query/processors/custom_function.py:    according to the SignatureValidator.
snuba/query/processors/custom_function.py:        self.__validator = SignatureValidator(param_types)
snuba/query/processors/handled_functions.py:from snuba.query.validation.signature import SignatureValidator
snuba/query/processors/handled_functions.py:        validator = SignatureValidator([])
```

Both custom_function and handled_function are logical query processors. That means the process method has access to the entity. From the entity they can get to the columnset. 
In general in the Snuba codebase this kind of refactoring is fairly safe as it is fully typed, so mypy can tell you immediately if you messed up. I would consider making that change in its own PR that is self contained and trivial to review, which should be pretty quick.

MeredithAnya commented:
> On that note, there is something entity specific that could be checked here: some entities allow custom functions using the `CustomFunction` processor. For example here in the transactions entity. https://github.com/getsentry/snuba/blob/master/snuba/datasets/entities/transactions.py#L144.
> 
> These processors allow Sentry to pass functions that don't exist in Clickhouse (`apdex(x, y)`) and the processor will convert it into the appropriate Clickhouse function. These processors are run after your validation, so there will be terms like `FunctionCall(None, function_name="apdex"...)` that will fail your validation because `apdex` is not whitelisted and the expression hasn't been converted to a Clickhouse function yet.

@evanh do you think it would be best to just add any customs functions to the global list for now?

evanh commented:
> @evanh do you think it would be best to just add any customs functions to the global list for now?

Yeah that makes the most sense I think.

MeredithAnya commented on snuba/query/parser/validation/functions.py:
```diff
@@ -38,39 +40,44 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
-            common_function_validators = (
-                entity_validators.keys() & default_validators.keys()
-            )
-            if common_function_validators:
```
also I got rid of this check because I don't really get why we need it. we don't have any entity_validators atm and it seems to make sense that if you had an entity specific one, you should use that over the default ?

fpacifici commented on snuba/query/validation/functions.py:
```diff
@@ -0,0 +1,35 @@
+        self.__enforce = enforce
+
+    def validate(
+        self, func_name: str, parameters: Sequence[Expression], schema: Any
```
Why is schema `Any` ?

fpacifici commented on snuba/query/validation/functions.py:
```diff
@@ -0,0 +1,35 @@
+            else:
+                logger.warning(f"Invalid function name: {func_name}", exc_info=True)
+
+    def __validate_impl(
+        self, func_name: str, parameters: Sequence[Expression], schema: Any
+    ) -> None:
+        if not is_valid_global_function(func_name):
+            raise InvalidFunctionCall(f"Invalid function name: {func_name}")
```
What is the reason to move the implementation in its own method? It does not seem to be ever used anywhere else and it contains two statements only.

fpacifici commented on snuba/query/parser/validation/functions.py:
```diff
@@ -25,6 +26,7 @@
     "like": SignatureValidator([Column({Array, String}), Any()]),
     "notLike": SignatureValidator([Column({Array, String}), Any()]),
 }
+global_validators: List[FunctionCallValidator] = [AllowedFunctionValidator()]
```
good, thanks

MeredithAnya commented on snuba/query/validation/functions.py:
```diff
@@ -0,0 +1,31 @@
+        if self.__enforce:
+            raise InvalidFunctionCall(f"Invalid function name: {func_name}")
+        else:
+            logger.warning(f"Invalid function name: {func_name}", exc_info=True)
```
@fpacifici @evanh in terms of using this to figure out what functions I might have missed from the list, I don't know if this is going to be too noisy if I miss a really simple one that's in like a very common query

fpacifici replied:
This would be ok but it would create a lot of different issues at first. 
I would advise to use a datadog metric and pass the function name as a tag.

evanh replied:
I second this.

fpacifici commented on snuba/query/parser/validation/functions.py:
```diff
@@ -103,34 +102,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
             raise InvalidExpressionException(
                 exp,
-                f"Illegal call to function {exp.function_name}: {str(exception)}",
+                f"Invalid function name {exp.function_name}: {str(exception)}",
```
?

fpacifici commented on snuba/query/parser/validation/functions.py:
```diff
@@ -103,34 +102,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
+
+            entity_validator = entity.get_function_call_validators().get(
+                exp.function_name
+            ) or default_validators.get(exp.function_name)
```
If you have a join query (thus with multiple entities) wouldn't you be applying the default validation multiple times if neither of the entity provides a validator for this function ?

fpacifici commented on snuba/query/validation/functions.py:
```diff
@@ -0,0 +1,31 @@
+logger = logging.getLogger(__name__)
+
+
+class AllowedFunctionValidator(FunctionCallValidator):
```
Could you please add a nit test for this class ?

fpacifici commented on tests/query/parser/validation/test_functions.py:
```diff
@@ -78,3 +82,21 @@ def test_functions(
+# If we were on pytest version 6.2.0 we could type this using pytest.LogCaptureFixture
+# https://stackoverflow.com/questions/68576519/what-is-the-type-hint-for-pytests-caplog-fixture
+def test_invalid_function_name(caplog: Any) -> None:
+    entity_return = MagicMock()
+    entity_return.return_value = {}
+    events_entity = get_entity(EntityKey.EVENTS)
+    setattr(events_entity, "get_function_call_validators", entity_return)
```
Why do you need to mock this? Events does not have any function validator.

fpacifici reviewed (commented):
Some comments in line but nothing major

evanh commented on snuba/query/parser/validation/functions.py:
```diff
@@ -59,32 +60,30 @@ class QueryEntityFinder(
 
-    def visit_join_clause(self, node: JoinClause[QueryEntity]) -> Optional[QueryEntity]:
+    def visit_join_clause(self, node: JoinClause[QueryEntity]) -> List[QueryEntity]:
         # Just returns one entity for now, later return both entities
```
💀

evanh commented on snuba/query/functions.py:
```diff
@@ -102,3 +102,51 @@
+    "tuple",
+}
+
+# Custom function names that are not in ClickHouse but need to be validated
+CUSTOM_FUNCTIONS = {"apdex", "failure_rate"}
```
Right now this is encoded as a custom function query processor on the entity. It's possible to iterate through the entity processors to find if an entity has any custom function processors though. Or we could move those processors into a specific field of the entity.

evanh commented on snuba/query/parser/validation/functions.py:
```diff
@@ -59,32 +60,30 @@ class QueryEntityFinder(
-    def _visit_simple_source(self, data_source: QueryEntity) -> QueryEntity:
-        return data_source
+    def _visit_simple_source(self, data_source: QueryEntity) -> List[QueryEntity]:
+        return [data_source]
```
Maybe this should return a Set? It's possible in a Join for the same entity to appear multiple times.

evanh commented on snuba/query/validation/functions.py:
```diff
@@ -0,0 +1,31 @@
+    """
+
+    def __init__(self, enforce: bool = False):
+        self.__enforce = enforce
```
I would make this based on a state setting. If you made the setting an int we could even ramp this up incrementally. That way you don't have to deploy to turn this on.

evanh reviewed (commented):
Looks pretty good. A few small comments.
